File: 3355.zero-array-transformation-i/solution.go
func isZeroArray(nums []int, queries [][]int) bool {
    changes := make([]int, len(nums) + 1)
    for _, q := range queries {
        changes[q[0]]++
        changes[q[1] + 1]--
    }

    maxValue := 0
    for i := range nums {
        maxValue += changes[i]
        if maxValue < nums[i] {
            return false
        }
    }
    return true
}

File: 3355.zero-array-transformation-i/solution.kt
class Solution {
    fun isZeroArray(nums: IntArray, queries: Array<IntArray>): Boolean {
        val changes = IntArray(nums.size + 1)
        queries.forEach { (l, r) ->
            changes[l]++
            changes[r + 1]--
        }
        return changes
            .scan(0) { acc, value -> acc + value }
            .drop(1)
            .zip(nums.asIterable())
            .all { (maxOps, num) -> maxOps >= num }        
    }
}